### ISSUE TITLE
About bn_update

### ISSUE THREAD

**Issue by zllrunning** [timgaripov/swa#10]
Hi, @timgaripov 
I know we should run one additional pass over the training data to get the running mean and running var. But why change momentum?
```
        momentum = b / (n + b)
        for module in momenta.keys():
            module.momentum = momentum
```

**Comment by zllrunning** [comment]
I did two experiments, one is to keep the momentum constant, and the other is to change the momentum as `momentum = b / (n + b)`. But keeping the momentum unchanged does not affect the experimental results.

**Comment by dahburj** [comment]
Just to clarify when you say it had no effect on the experimental results are you speaking in terms of just accuracy/loss or was there actually no change whatsoever across the board? (Convergence time, processing power, etc.)

**Comment by zllrunning** [comment]
@dahburj  You can see my training log. 
Change the momentum : momentum = b / (n + b)
```
----  --------  ---------  --------  ---------  --------  -------------  ------------  --------
  ep        lr    tr_loss    tr_acc  te_loss    te_acc    swa_te_loss    swa_te_acc        time
----  --------  ---------  --------  ---------  --------  -------------  ------------  --------
 281    0.0100     0.2255   93.8620                                                     13.2292
 282    0.0100     0.2409   93.6000                                                     13.3089
 283    0.0100     0.2434   93.4260                                                     12.9591
 284    0.0100     0.2346   93.5140                                                     12.8891
 285    0.0100     0.2415   93.5600     1.6920   67.1900         1.1433       76.3700   19.1549
 286    0.0100     0.2332   93.7240                                                     12.9103
 287    0.0100     0.2387   93.5100                                                     13.0078
 288    0.0100     0.2403   93.6000                                                     13.5074
 289    0.0100     0.2431   93.4840                                                     12.9497
 290    0.0100     0.2236   93.9840     1.8050   65.5200         1.1427       76.4600   19.0503
 291    0.0100     0.2317   93.6900                                                     12.9085
 292    0.0100     0.2400   93.5200                                                     13.0002
 293    0.0100     0.2375   93.6740                                                     12.9674
 294    0.0100     0.2298   93.7340                                                     12.9769
 295    0.0100     0.2242   94.0100     1.7459   66.6500         1.1440       76.4600   19.0024
 296    0.0100     0.2381   93.6120                                                     12.8901
 297    0.0100     0.2344   93.5940                                                     12.9130
 298    0.0100     0.2391   93.5680                                                     12.9390
 299    0.0100     0.2355   93.5100                                                     12.8852
 300    0.0100     0.2238   93.9480     1.7326   67.5900         1.1437       76.4500   19.0932
```

Keep  the momentum constant:
```
----  --------  ---------  --------  ---------  --------  -------------  ------------  --------
  ep        lr    tr_loss    tr_acc  te_loss    te_acc    swa_te_loss    swa_te_acc        time
----  --------  ---------  --------  ---------  --------  -------------  ------------  --------
 281    0.0100     0.2402   93.4920                                                     13.1742
 282    0.0100     0.2339   93.6080                                                     13.5598
 283    0.0100     0.2350   93.6940                                                     13.0823
 284    0.0100     0.2445   93.4160                                                     13.0438
 285    0.0100     0.2388   93.4420     1.6882   67.5300         1.1261       76.7300   19.2877
 286    0.0100     0.2378   93.5900                                                     13.0436
 287    0.0100     0.2484   93.3440                                                     13.2893
 288    0.0100     0.2370   93.5600                                                     13.0326
 289    0.0100     0.2275   93.8540                                                     12.8917
 290    0.0100     0.2326   93.7020     1.7479   66.5100         1.1291       76.7700   18.8673
 291    0.0100     0.2317   93.8820                                                     12.8628
 292    0.0100     0.2263   93.7160                                                     12.8964
 293    0.0100     0.2346   93.6900                                                     12.8455
 294    0.0100     0.2407   93.5280                                                     12.9580
 295    0.0100     0.2334   93.5620     1.6953   67.0300         1.1282       76.5500   18.8932
 296    0.0100     0.2453   93.4280                                                     13.1283
 297    0.0100     0.2286   93.7940                                                     13.2678
 298    0.0100     0.2372   93.6200                                                     12.9918
 299    0.0100     0.2281   93.7860                                                     12.9640
 300    0.0100     0.2521   93.2440     1.7465   65.8100         1.1330       76.6300   19.0337

```

**Comment by dahburj** [comment]
That is interesting, I would have thought keeping the momentum the same would make the SWA model suffer and/or (at least) converge slower since we're not accounting for the number of SWA models we trained. Did you run this on the CIFAR10/0 data set?

**Comment by izmailovpavel** [comment]
Hi, sorry for late response. Batchnorm is trying to estimate the activation statistics over many epochs, and that's why it computes a running average (with a certain momentum) of activation statistics. We, however, only run one epoch to update batchnorm parameters, and in for this reason we just compute an equally-weighted average of these statistics. As you noticed, this choice doesn't make a huge difference, if we have a large enough dataset.
